### ISSUE TITLE
技巧-Pytorch

### ISSUE THREAD

**Issue by LarT2P** [lartpang/Machine-Deep-Learning#39]
# 损失函数整理

https://blog.csdn.net/zhangxb35/article/details/72464152?utm_source=itdadao&utm_medium=referral

**Comment by LarT2P** [comment]
# 保存/载入模型

```python
# 存储一个epoch后的模型(权重和偏置项), 以便后期使用
filename = ('%s/feature-current.pth' % check_root_feature)
torch.save(model.state_dict(), filename)

# 存储优化器状态
filename_opti = ('%s/opti-current.pth' % check_root_opti)
torch.save(optimizer_feature.state_dict(), filename_opti)

# 载入上一次的训练结果(权重和偏置项), 进一步的训练
model.load_state_dict(
    torch.load(check_root_feature + '/feature-current.pth')
)
# 载入优化器状态
optimizer_feature.load_state_dict(
    torch.load(check_root_opti + '/opti-current.pth')
)
```

```python
fileroot = ('%s/feature-current.pth' % check_root_feature)
# 基于torch.save(model.state_dict(), filename)存储方法的对应的恢复方法
model.load_state_dict(torch.load(fileroot))
```

**Comment by LarT2P** [comment]
# 展示数据

```python
# Have a look at data
inputs, classes = next(iter(dataloaders['train']))
out = torchvision.utils.make_grid(inputs)
imshow(out, title=[class_names[x] for x in classes])
```

**Comment by LarT2P** [comment]
# 反卷积计算

![image](https://user-images.githubusercontent.com/16298490/50755881-62e34780-1295-11e9-9b43-2acfcb58611c.png)

**Comment by LarT2P** [comment]
# 图像相关python库

* Numpy：主要作为图像数据的中转
* Skimage：速度较慢，主要是caffe测试的时候默认的caffe.io.load使用，不建议使用（Skimage，读取RGB，通道HWC，范围[0,1]）
* 主要的图像处理包：
	* CV2：opencv的python版本，应用较广
	* PIL：Python自带的图像处理库，但功能较简单
	* Pillow: 出发点在于PIL只支持python2.7，添加了一些新特性
	* Pillow-SIMD：Pillow增强版本，约有5倍的增益，比CV2还要快；安装方法：
		```sh
		$ pip uninstall -y pillow 
		$ CC="cc -mavx2" pip install -U --force-reinstall pillow-simd
		```
	* Matplotlib：一般使用matplot.pyplot方法进行画图；

## 图像处理

### Numpy

Numpy对多维矩阵A的操作一般有：

* A.shape #HWC
* type(A) #numpy.array
* A.dtype() #uint8, float…
* np.min(A), np.max(A) #最值

## CV2

* 读取BGR，通道HWC，范围[0,255] ，类型uint8; 
* 图像类型numpy.ndarray；

## PIL，Pillow, Pillow-SIMD

* 读取RGB，通道HWC，范围[0,255]，类型uint8；
* 图像类型PngImageFile （np.array, Image.fromarray直接与numpy互相转换）
* 有.mode方法---rgb信息

## Matplotlib

* 读取RGB，通道HWC，范围[0,1] ，类型float；
* 图像类型numpy.ndarray

## Skimage

* 读取RGB，通道HWC，范围[0,255]，类型uint8；
* 图像类型numpy.ndarray 
* 有.mode方法---rgb信息
* 比较特殊，读取的时候image= io.imread('test.jpg',as_grey=False)；
* 彩图是uint8，[0,255]；灰度图float，[0,1]；
* 彩图resize变成float，[0,1]；
* 较混乱，不适用。。。

## 读取

### CV2

* 默认读取为三通道彩图，可通过参数更改为单通道灰度图
	* `img = cv2.imread('examples.png')`
	* `img_gray = cv2.imread('examples.png', 0)`

### PIL，Pillow, Pillow-SIMD

`img = Image.open('examples.png')`

### Matplotlib

`img = plt.imread('examples.png')`

## 显示

### Matplotlib

```python
# 最主要目的是用来绘图: 将numpy数组格式的RGB图像显示；float类型的图像，范围0-1；如果是uint8图像，范围是0-255；
img = plt.imread('examples.png')
plt.imshow(img)
plt.show()
```

### CV2

```python
img = cv2.imread('examples.png')
plt.imshow(img[..., -1::-1]) # 因为opencv读取进来的是bgr顺序，而imshow需要的是rgb顺序，因此需要先反过来,也可以plt.imshow(img[:,:,::-1])
plt.show()
```

### PIL

```python
#可直接打开
plt.imshow(Image.open('examples.png')) # 实际上plt.imshow可以直接显示PIL格式图像
plt.show()
#转换为需要的numpy格式打开
img_gray = img.convert('L') #转换成灰度图像
img = np.array(img)
img_gray = np.array(img_gray)
plt.imshow(img) # or plt.imshow(img / 255.0)5
plt.show()
plt.imshow(img_gray, cmap=plt.gray()) # 显示灰度图要设置cmap参数
plt.show()
```
## 转换

主要是通过numpy的transpose操作，修正RGB，BGR；

例如：

* a是rgb图像，那么 a[::-1]，a[:,::-1]，a[:,:,::-1]分别是X轴的镜像，Y轴的镜像，BGR转换为RGB；

### opencv相关图像操作

```python
img_gray = cv2.cvtColor(img,
cv2.COLOR_BGR2GRAY) # BGR转灰度

img_bgr = cv2.cvtColor(img_gray,
cv2.COLOR_GRAY2BGR) # 灰度转BRG

img_rgb = cv2.cvtColor(img_gray,
cv2.COLOR_GRAY2RGB) # 灰度转RGB

b,g,r = cv2.split(img) #bgr图像分离三个通道
img2 = cv2.merge([r,g,b]) #merge成rgb图像

### PIL相关图像操作

```python
img = Image.open('examples.png')
img_gray = image.convert(‘L’)
img_color = img_gray.convert(‘RGB’)

### PIL与numpy格式转换操作

```python
numpy.asarray()
Image.fromarray()
```

如果是pil转opencv，记得需要通过copy命令得到的才可以进行cv2操作，不然会有bug。切记.

## 尺寸

```python
# PIL类型，尺寸信息，通过.size方法，得到WH
print image.size #width height
# 同样进行resize操作，顺序也是wh；
img2_resize = img2.resize((960,540))
img2_resize.save('test1.jpg')
# 而如果使用cv2操作，顺序也是wh；
img3_resize = cv2.resize(img3, (960,540))
cv2.imwrite('test2.jpg', img3_resize)
# 但如果放在numpy里面，调用shape方法，得到的是HWC；
```

## 保存

### PIL

```python
#直接save方法
img = Image.open('examples.png')
img.save('examples2.png')
img_gray = img.convert('L')
img_gray.save('examples_gray.png') # 不管是灰度还是彩色，直接用save函数保存就可以，但注意，只有PIL格式的图片能够用save函数

### CV2

```python
import cv2
img = cv2.imread('examples.png') # 这是BGR图片
cv2.imwrite('examples2.png', img) # 这里也应该用BGR图片保存，这里要非常注意，因为用pylab或PIL读入的图片都是RGB的，如果要用opencv存图片就必须做一个转换
img_gray = cv2.cvtColor(img,
cv2.COLOR_BGR2GRAY)
cv2.imwrite('examples_gray.png', img_gray)
```

**Comment by LarT2P** [comment]
# 显存优化

+ Optimizing PyTorch training code: <https://www.sagivtech.com/2017/09/19/optimizing-pytorch-training-code/>
	- 及时删除中间变量 `del`
	- 避免不必要的数据搬移
	- 使用固定内存，并使用 `async = True` 来并行化数据传输和GPU数字运算

**Comment by LarT2P** [comment]
## pytorch加速技巧

https://zhuanlan.zhihu.com/p/39752167

**Comment by LarT2P** [comment]
## TensorBoardX

```python
# demo.py

import torch
import torchvision.utils as vutils
import numpy as np
import torchvision.models as models
from torchvision import datasets
from tensorboardX import SummaryWriter

resnet18 = models.resnet18(False)
writer = SummaryWriter()
sample_rate = 44100
freqs = [262, 294, 330, 349, 392, 440, 440, 440, 440, 440, 440]

for n_iter in range(100):

    dummy_s1 = torch.rand(1)
    dummy_s2 = torch.rand(1)
    # data grouping by `slash`
    writer.add_scalar('data/scalar1', dummy_s1[0], n_iter)
    writer.add_scalar('data/scalar2', dummy_s2[0], n_iter)

    writer.add_scalars('data/scalar_group', {'xsinx': n_iter * np.sin(n_iter),
                                             'xcosx': n_iter * np.cos(n_iter),
                                             'arctanx': np.arctan(n_iter)}, n_iter)

    dummy_img = torch.rand(32, 3, 64, 64)  # output from network
    if n_iter % 10 == 0:
        x = vutils.make_grid(dummy_img, normalize=True, scale_each=True)
        writer.add_image('Image', x, n_iter)

        dummy_audio = torch.zeros(sample_rate * 2)
        for i in range(x.size(0)):
            # amplitude of sound should in [-1, 1]
            dummy_audio[i] = np.cos(freqs[n_iter // 10] * np.pi * float(i) / float(sample_rate))
        writer.add_audio('myAudio', dummy_audio, n_iter, sample_rate=sample_rate)

        writer.add_text('Text', 'text logged at step:' + str(n_iter), n_iter)

        for name, param in resnet18.named_parameters():
            writer.add_histogram(name, param.clone().cpu().data.numpy(), n_iter)

        # needs tensorboard 0.4RC or later
        writer.add_pr_curve('xoxo', np.random.randint(2, size=100), np.random.rand(100), n_iter)

dataset = datasets.MNIST('mnist', train=False, download=True)
images = dataset.test_data[:100].float()
label = dataset.test_labels[:100]

features = images.view(100, 784)
writer.add_embedding(features, metadata=label, label_img=images.unsqueeze(1))

# export scalar data to JSON for external processing
writer.export_scalars_to_json("./all_scalars.json")
writer.close()
```

**Comment by LarT2P** [comment]
## 梯度剪裁

```python
optimizer.zero_grad()        
loss, hidden = model(data, hidden, targets)
loss.backward()
 
torch.nn.utils.clip_grad_norm(model.parameters(), args.clip)
optimizer.step()
```

既然在BP过程中会产生梯度消失/爆炸（就是偏导无限接近0，导致长时记忆无法更新），那么最简单粗暴的方法，设定阈值，当梯度小于/大于阈值时，更新的梯度为阈值.

![image](https://user-images.githubusercontent.com/16298490/54998068-2ad3e100-5008-11e9-8c54-ef937e5a473a.png)

**Comment by LarT2P** [comment]
## 参数初始化

```
 # 1. 根据网络层的不同定义不同的初始化方式     
def weight_init(m):
    if isinstance(m, nn.Linear):
        nn.init.xavier_normal_(m.weight)
        nn.init.constant_(m.bias, 0)
    # 也可以判断是否为conv2d，使用相应的初始化方式 
    elif isinstance(m, nn.Conv2d):
        nn.init.kaiming_normal_(m.weight, mode='fan_out', nonlinearity='relu')
     # 是否为批归一化层
    elif isinstance(m, nn.BatchNorm2d):
        nn.init.constant_(m.weight, 1)
        nn.init.constant_(m.bias, 0)
# 2. 初始化网络结构        
model = Net(in_dim, n_hidden_1, n_hidden_2, out_dim)
# 3. 将weight_init应用在子模块上
model.apply(weight_init)
```

### 参考链接

* https://blog.csdn.net/dss_dssssd/article/details/83990511

**Comment by LarT2P** [comment]
## Unfold & Fold

```pytorch
temp = torch.randint(0, 9, (1, 2, 2, 2), dtype=torch.float32)
temp
Out[12]: 
tensor([[[[2., 7.],
          [1., 3.]],
         [[2., 0.],
          [1., 1.]]]])
F.unfold(temp, (1, 2), 1, 0, 1)
Out[13]: 
tensor([[[2., 1.],
         [7., 3.],
         [2., 1.],
         [0., 1.]]])
F.unfold(temp, (2, 1), 1, 0, 1)
Out[14]: 
tensor([[[2., 7.],
         [1., 3.],
         [2., 0.],
         [1., 1.]]])
F.unfold(temp, (2, 2), 1, 0, 1)
Out[15]: 
tensor([[[2.],
         [7.],
         [1.],
         [3.],
         [2.],
         [0.],
         [1.],
         [1.]]])
```

## view

```python
temp.view(1, 2*2*2, 1, 1)
Out[16]: 
tensor([[[[2.]],
         [[7.]],
         [[1.]],
         [[3.]],
         [[2.]],
         [[0.]],
         [[1.]],
         [[1.]]]])
temp.view(1, 2*2, 2, 1)
Out[17]: 
tensor([[[[2.],
          [7.]],
         [[1.],
          [3.]],
         [[2.],
          [0.]],
         [[1.],
          [1.]]]])
temp.view(1, 2*2, 1, 2)
Out[18]: 
tensor([[[[2., 7.]],
         [[1., 3.]],
         [[2., 0.]],
         [[1., 1.]]]])
```

**Comment by LarT2P** [comment]
## 矩阵乘法

torch.addmm
对矩阵mat1和mat2进行矩阵乘操作(用@表示)。矩阵mat加到最终结果。

```python
out=(beta∗M)+(alpha∗mat1@mat2)
torch.addmm(beta=1, mat, alpha=1, mat1, mat2, out=None)
```

torch.mm
对矩阵mat1和mat2进行相乘。

```python
torch.mm(mat1, mat2, out=None)
```

torch.bmm
对存储在两个批batch1和batch2内的矩阵进行批矩阵乘操作。torch.bmm(batch1, batch2, out=None)

用法：

```python
>>> batch1 = torch.randn(10, 3, 4)
>>> batch2 = torch.randn(10, 4, 5)
>>> res = torch.bmm(batch1, batch2)
>>> res.size()
torch.Size([10, 3, 5])
```

**Comment by lartpang** [comment]
## Shuffle Channels

主要用在了shufflenet中，是一种修改通道顺序的操作，从各组中抽取特定通道进行组合。

```python
    def shuffle_channels(x, groups):
        '''Channel shuffle: [N,C,H,W] -> [N,g,C/g,H,W] -> [N,C/g,g,H,W] -> [N,C,H,W]'''
        '''一共C个channel要分成g组混合的channel，先把C reshape成(g, C/g)的形状，然后转置成(C/g, g)最后平坦成C组channel'''
        N, C, H, W = x.size()
        return x.view(N, groups, C // groups, H, W).permute(0, 2, 1, 3, 4).contiguous().view(N, C, H, W)  # 因为x之前view过了，他的内存不连续了，需要contiguous来规整一下
--------------------- 
作者：急流勇进 
来源：CSDN 
原文：https://blog.csdn.net/weixin_44538273/article/details/88856239 
版权声明：本文为博主原创文章，转载请附上博文链接！
```

**Comment by lartpang** [comment]
## Subpixel Conv的像素处理操作

https://github.com/pytorch/examples/blob/1de2ff9338bacaaffa123d03ce53d7522d5dcc2e/super_resolution/model.py#L15

```python
nn.PixelShuffle(upscale_factor)
```

该操作实现了下图的操作，只是重排了特征，没有其他操作：

![image](https://user-images.githubusercontent.com/26847524/60392226-091b5600-9b31-11e9-8cbb-10fa50ede2ad.png)
